### ISSUE TITLE
List halt on sorted(by:)

### ISSUE THREAD

**Issue by YuantongL** [aws-amplify/amplify-ios#1526]
### Describe the bug

`List` conforms to `Swift.collection`, so we can apply `.sorted(by:)` to it, as my example shows, this will immediate halt the app without any indication.

### Steps To Reproduce
`conversations` is of type `List<UserConversation>?`
```swift
let sortedConversations = conversations?.sorted(by: { lhs, rhs in
  guard let lhsTime = lhs.updatedAt, let rhsTime = rhs.updatedAt else {
    return true
  }
  return lhsTime < rhsTime
})
```
The program will halt on `sorted(by:)`, I think it is right after `List.elements` is referenced.

### Expected behavior

It sorts the element in list

### Amplify Framework Version

1.14.0

### Amplify Categories

API

### Dependency manager

Swift PM

### Swift version

5.5.1

### CLI version

6.3.1

### Xcode version

13.1

### Is this a regression?

No

### Device

iPhone 13 pro max simulator

### iOS Version

15.0

**Comment by atierian** [comment]
Hi @YuantongL, thanks for opening this issue.

I was unable to reproduce the issue you're experiencing.
Here is the code used to try to reproduce this and the output. 

```swift
/// Simplified `UserConversation` Model
struct UserConversation: Model {
    var id: Identifier
    var updatedAt: Temporal.DateTime?

    init(_ timeString: String?) {
        id = UUID().uuidString
        updatedAt = try? timeString.map(Temporal.DateTime.init(iso8601String:))
    }
}

/// Formatted debug output for easier debugging here
extension UserConversation: CustomDebugStringConvertible {
    var debugDescription: String {
        """
        ID: \(id)
        updatedAt: \(updatedAt?.iso8601FormattedString(format: .full) ?? "--empty--")
        ~~~~~~~~
        """
    }
}

struct Issue {
    /// List of UserConversations to be sorted
    let conversations: List<UserConversation>? = List(
        elements: [
            UserConversation(nil),
            UserConversation("2021-12-09T04:15:00.000+0000"),
            UserConversation("2021-12-09T07:00:00.000+0000"),
            UserConversation("2021-12-09T07:00:00.001+0000"),
            UserConversation("2021-12-09T07:00:00.013+0000")
        ]
    )

    /// Test the reported issue by shuffling conversations, then attempting to sort.
    func test() {
        let conversations = conversations?.shuffled()
        conversations?.forEach { print($0) }
        /* OUTPUT:
          ID: 6CBB2EE1-71AE-4886-A733-5C7B1B079A38
          updatedAt: 2021-12-09T07:00:00.000Z
          ~~~~~~~~
          ID: 5271D36C-0B87-46B5-9EA6-672FC23D8602
          updatedAt: 2021-12-09T07:00:00.013Z
          ~~~~~~~~
          ID: 27C11373-FE7D-4D7D-86EA-278AD94F0C0A
          updatedAt: 2021-12-09T07:00:00.001Z
          ~~~~~~~~
          ID: 510D3C10-9A51-40A4-8747-A2B4D5CAA379
          updatedAt: 2021-12-09T04:15:00.000Z
          ~~~~~~~~
          ID: 21745851-70AC-4269-A583-9104C9F8E727
          updatedAt: --empty--
          ~~~~~~~~
         */

        let sortedConversation = conversations?.sorted(by: { lhs, rhs in
            guard let lhsTime = lhs.updatedAt, let rhsTime = rhs.updatedAt else { return true }
            return lhsTime < rhsTime
        })

        sortedConversation?.forEach { print($0) }
        /* OUTPUT:
          ID: 21745851-70AC-4269-A583-9104C9F8E727
          updatedAt: --empty--
          ~~~~~~~~
          ID: 510D3C10-9A51-40A4-8747-A2B4D5CAA379
          updatedAt: 2021-12-09T04:15:00.000Z
          ~~~~~~~~
          ID: 6CBB2EE1-71AE-4886-A733-5C7B1B079A38
          updatedAt: 2021-12-09T07:00:00.000Z
          ~~~~~~~~
          ID: 27C11373-FE7D-4D7D-86EA-278AD94F0C0A
          updatedAt: 2021-12-09T07:00:00.001Z
          ~~~~~~~~
          ID: 5271D36C-0B87-46B5-9EA6-672FC23D8602
          updatedAt: 2021-12-09T07:00:00.013Z
          ~~~~~~~~
         */
    }
}

// Test
let issue = Issue()
issue.test()
```


Can you please elaborate on what you mean by: _this will immediate halt the app without any indication_?
If you can provide a simple reproducible case, that would be ideal.

Some data points that will help us diagnose this: 
- Does the app freeze and/or crash?
- Is any code executed after the sort?
- Is there any debug console output?
- Does the memory usage and/or CPU usage of the app spike when sorting?
- Anything else you think may be helpful.

**Comment by YuantongL** [comment]
@atierian Thank you for your response, unfortunately I can't reproduce it now as I changed a lot of places in my project and I'm unable to trace back.

But from what I remember, it was a `List` which `AppSyncListProvider` was the provider, and when you sort when `list.loadState == .notLoaded` this happens.

I stepped into the code and saw it halt around [this line](https://github.com/aws-amplify/amplify-ios/blob/02ac2ba5c417883ea90f3d59b3cd28e63d2a1d39/AmplifyPlugins/API/AWSAPICategoryPlugin/Core/AppSyncListProvider.swift#L107) in `AppSyncListProvider`, it looks like a deadlock created by these `DispatchSemaphore` somehow.

I think you'll have to try use a `AppSyncListProvider` to reproduce it, I will also keep an eye on it as well. Thanks!
